File: benchmarks/docker/Dockerfile.go
# Multi-stage build para o bench harness Go (pgx/v5).
#
# Stage 1 (builder, golang:1.23): baixa deps, compila estaticamente
# CGO_ENABLED=0 para produzir binário portátil.
#
# Stage 2 (runtime, debian:stable-slim): só o binário + ca-certs.
# Imagem final ~25MB.

FROM golang:1.23 AS builder

WORKDIR /src

COPY bench/go/go.mod bench/go/go.sum ./
RUN go mod download

COPY bench/go/ .

ENV CGO_ENABLED=0
RUN go build -o /out/bench-go -ldflags="-s -w" .

FROM debian:stable-slim

RUN apt-get update && apt-get install -y --no-install-recommends \
        ca-certificates \
        postgresql-client \
    && rm -rf /var/lib/apt/lists/*

WORKDIR /app
COPY --from=builder /out/bench-go /app/bench-go

VOLUME ["/outputs"]

ENTRYPOINT ["/app/bench-go"]
CMD ["--workers", "1", "--conns-per-worker", "1", "--tx-count", "5000", "--reps", "3"]

File: benchmarks/docker/Dockerfile.java
# Multi-stage build for Java bench (pgjdbc).
# Stage 1 (maven:3.9-eclipse-temurin-21 builder): builds shaded fat jar.
# Stage 2 (eclipse-temurin:21-jre-alpine runtime): only JRE + the jar.
# Final image ~200MB (Alpine JRE + driver jar).

FROM maven:3.9-eclipse-temurin-21 AS builder

WORKDIR /src

COPY benchmarks/drivers/java/pom.xml .
RUN mvn -B -q dependency:go-offline

COPY benchmarks/drivers/java/src ./src
RUN mvn -B -q package -DskipTests

FROM eclipse-temurin:21-jre-alpine

RUN apk add --no-cache postgresql16-client

WORKDIR /app
COPY --from=builder /src/target/bench-java.jar /app/bench-java.jar

VOLUME ["/outputs"]

# G1 GC + headless server-class JIT defaults. -XX:+UseG1GC is implicit
# in Java 21 but we set explicitly for reproducibility.
ENTRYPOINT ["java", "-XX:+UseG1GC", "-Xms256m", "-Xmx2g", "-jar", "/app/bench-java.jar"]
CMD ["--workers", "1", "--conns-per-worker", "1", "--tx-count", "5000", "--reps", "3"]
